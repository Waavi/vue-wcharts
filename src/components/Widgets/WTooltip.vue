<template>
    <div
        v-if="selected"
        ref="tooltip"
        class="WTooltip"
        :class="{ visible }"
        :style="{ top: y, left: x }"
    >
        <slot v-bind="selected">
            <div class="Wrapper">
                <span
                    v-if="selected.label"
                    class="Title"
                >{{ selected.label }}</span>
                <div
                    v-for="(value, index) in selected.value"
                    :key="index"
                    class="Value"
                >
                    <slot
                        name="bullet"
                        v-bind="selected"
                    >
                        <div
                            class="Bullet"
                            :style="{ background: value.color }"
                        />
                    </slot>
                    <span>{{ value.value }}</span>
                </div>
            </div>
        </slot>
    </div>
</template>

<script>
import VueTypes from 'vue-types'
import { toPx } from './utils'

export default {
    name: 'WTooltip',
    type: 'plugins',
    inject: ['Chart'],
    props: {
        id: VueTypes.string,
        gap: VueTypes.number.def(10),
    },
    data () {
        return {
            selected: null,
            visible: false,
            x: 0,
            y: 0,
        }
    },
    computed: {
        // Bullet style
        bulletStyle () {
            if (!this.selected) return {}

            return {
                background: this.selected.color,
            }
        },
    },
    watch: {
        'Chart.active': function watchActive ({ el, event }) {
            // Reset tooltip if not has selected
            if (!el) {
                this.reset()
                return
            }

            // Set selected
            this.selected = el

            // Calc tooltip pos after rendering html
            this.$nextTick(() => {
                this.calcPos(event)
            })
        },
    },
    methods: {
        // Show tooltip
        show () {
            this.visible = true
        },
        // Hide tooltip
        hide () {
            this.visible = false
        },
        // Reset
        reset () {
            this.hide()
            this.selected = null
        },
        // Set pos tooltip
        calcPos (event) {
            const { layerX, layerY } = event
            const { viewWidth, height: viewHeight } = this.Chart
            const { offsetLeft = 0, offsetTop = 0 } = this.selected
            // Add space between point of tooltip
            let x = layerX + this.gap + offsetLeft
            let y = layerY + this.gap + offsetTop
            // Get size of tooltip
            const { clientWidth: elWidth, clientHeight: elHeight } = this.$refs.tooltip || {}

            // Move to left, if tooltip does not fit on the right
            if (viewWidth <= x + elWidth) x = x - elWidth - this.gap
            // Move to top, if tooltip does not fit on the bottom
            if (viewHeight <= y + elHeight) y = y - elHeight - this.gap

            // Set coords
            this.x = toPx(x)
            this.y = toPx(y)
            // Show tooltip
            this.$nextTick(this.show)
        },
    },
}
</script>

<style scoped lang="scss">
.WTooltip {
    position: absolute;
    display: flex;
    padding: .5rem;
    border-radius: 4px;
    color: #FFF;
    background: #333;
    font-size: 14px;
    opacity: 0;
    z-index: 1;
    transition: opacity .15s ease;

    &.visible {
        opacity: 1;
    }
}

.Wrapper {
    display: flex;
    flex-direction: column;
}

.Title {
    margin-bottom: 5px;
}

.Bullet {
    display: inline-block;
    height: 10px;
    width: 10px;
    border-radius: 50%;
    margin-right: .5rem;
}
</style>
