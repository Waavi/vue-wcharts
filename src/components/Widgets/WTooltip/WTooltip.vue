<template>
    <div
        v-if="selected"
        ref="tooltip"
        :style="{
            ...stylesCmp,
            ...(visible ? visibleStylesCmp : {}),
            top: y,
            left: x
        }"
    >
        <slot v-bind="selected">
            <div :style="wrapperStylesCmp">
                <slot
                    name="label"
                    :label="selected.label"
                >
                    <span
                        v-if="selected.label"
                        :style="titleStylesCmp"
                    >{{ selected.label }}</span>
                </slot>

                <div
                    v-for="(value, index) in selected.value"
                    :key="index"
                >
                    <slot
                        name="bullet"
                        v-bind="value"
                    >
                        <WBullet :styles="{ borderColor: value.color }" />
                    </slot>

                    <slot
                        name="value"
                        v-bind="value"
                    >
                        <span>{{ value.value }}</span>
                    </slot>
                </div>
            </div>
        </slot>
    </div>
</template>

<script>
import VueTypes from 'vue-types'
import WBullet from '../WBullet/WBullet.vue'
import { toPx } from '../utils'
import themeMixin from '../../../mixins/theme'

export default {
    name: 'WTooltip',
    type: 'plugins',
    inject: ['Chart'],
    components: {
        WBullet,
    },
    mixins: [themeMixin],
    props: {
        id: VueTypes.string,
        gap: VueTypes.number.def(10),
        styles: VueTypes.object,
        visibleStyles: VueTypes.object,
        titleStyles: VueTypes.object,
        wrapperStyles: VueTypes.object,
    },
    data () {
        return {
            selected: null,
            visible: false,
            x: 0,
            y: 0,
        }
    },
    computed: {
        // Styles
        stylesCmp () {
            return {
                ...this.themeStyles.styles,
                ...this.styles,
            }
        },
        // Visible Styles
        visibleStylesCmp () {
            return {
                ...this.themeStyles.visible,
                ...this.visibleStyles,
            }
        },
        // Wrapper Styles
        wrapperStylesCmp () {
            return {
                ...this.themeStyles.wrapper,
                ...this.wrapperStyles,
            }
        },
        // Title Styles
        titleStylesCmp () {
            return {
                ...this.themeStyles.title,
                ...this.titleStyles,
            }
        },
        // Bullet style
        bulletStyle () {
            if (!this.selected) return {}

            return {
                background: this.selected.color,
            }
        },
    },
    watch: {
        'Chart.active': function watchActive ({ el, event }) {
            // Reset tooltip if not has selected
            if (!el) {
                this.reset()
                return
            }

            // Set selected
            this.selected = el

            // Calc tooltip pos after rendering html
            this.$nextTick(() => {
                this.calcPos(event)
            })
        },
    },
    methods: {
        // Show tooltip
        show () {
            this.visible = true
        },
        // Hide tooltip
        hide () {
            this.visible = false
        },
        // Reset
        reset () {
            this.hide()
            this.selected = null
        },
        // Set pos tooltip
        calcPos (event) {
            const { layerX, layerY } = event
            const { viewWidth, height: viewHeight } = this.Chart
            const { offsetLeft = 0, offsetTop = 0 } = this.selected
            // Add space between point of tooltip
            let x = layerX + this.gap + offsetLeft
            let y = layerY + this.gap + offsetTop
            // Get size of tooltip
            const { clientWidth: elWidth, clientHeight: elHeight } = this.$refs.tooltip || {}

            // Move to left, if tooltip does not fit on the right
            if (viewWidth <= x + elWidth) x = x - elWidth - this.gap
            // Move to top, if tooltip does not fit on the bottom
            if (viewHeight <= y + elHeight) y = y - elHeight - this.gap

            // Set coords
            this.x = toPx(x)
            this.y = toPx(y)
            // Show tooltip
            this.$nextTick(this.show)
        },
    },
}
</script>
