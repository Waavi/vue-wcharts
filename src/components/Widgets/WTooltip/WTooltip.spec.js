import { shallowMount } from '@vue/test-utils'
import WTooltip from './WTooltip.vue'

describe('Widgets/WTooltip', () => {
    const propsData = {
        id: 'test',
        gap: 0,
    }

    const provide = {
        Chart: {
            legends: ['One Bar', 'Two Bar'],
            space: [20, 20, 20, 20],
            activeElements: [0, 1],
            viewWidth: 600,
            viewHeight: 400,
        },
    }

    const defaultConfig = {
        propsData,
        provide,
    }

    it(`Should be render correctly`, () => {
        const wrapper = shallowMount(WTooltip, defaultConfig)
        expect(wrapper.html()).toMatchSnapshot()
    })
})
