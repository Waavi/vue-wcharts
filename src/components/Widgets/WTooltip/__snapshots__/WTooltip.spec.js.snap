// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Widgets/WTooltip Methods Should be set x and y prop after call calcPos method 1`] = `
<div style="position: absolute; display: flex; padding: .5rem; border-radius: 4px; color: rgb(0, 0, 0); background: white; box-shadow: 0 2px 10px 0 rgba(0, 0, 0, 0.15); font-size: 14px; opacity: 0; z-index: 1; transition: opacity .25s ease; top: 10px; left: 123px;">
  <div style="display: flex; flex-direction: column;">
    <!---->
  </div>
</div>
`;

exports[`Widgets/WTooltip Should be render correctly 1`] = `
<div style="position: absolute; display: flex; padding: .5rem; border-radius: 4px; color: rgb(0, 0, 0); background: white; box-shadow: 0 2px 10px 0 rgba(0, 0, 0, 0.15); font-size: 14px; opacity: 0; z-index: 1; transition: opacity .25s ease; top: 0px; left: 0px;">
  <div style="display: flex; flex-direction: column;">
    <!---->
  </div>
</div>
`;

exports[`Widgets/WTooltip Should be render correctly with label 1`] = `
<div style="position: absolute; display: flex; padding: .5rem; border-radius: 4px; color: rgb(0, 0, 0); background: white; box-shadow: 0 2px 10px 0 rgba(0, 0, 0, 0.15); font-size: 14px; opacity: 0; z-index: 1; transition: opacity .25s ease; top: 0px; left: 0px;">
  <div style="display: flex; flex-direction: column;"><span style="margin-bottom: 5px;">test</span> </div>
</div>
`;

exports[`Widgets/WTooltip Should be render correctly with values 1`] = `
<div style="position: absolute; display: flex; padding: .5rem; border-radius: 4px; color: rgb(0, 0, 0); background: white; box-shadow: 0 2px 10px 0 rgba(0, 0, 0, 0.15); font-size: 14px; opacity: 0; z-index: 1; transition: opacity .25s ease; top: 0px; left: 0px;">
  <div style="display: flex; flex-direction: column;">
    <!---->
    <div>
      <wbullet-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#5C6BC0,#FB499D" styles="[object Object]"></wbullet-stub> <span>1</span>
    </div>
    <div>
      <wbullet-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#5C6BC0,#FB499D" styles="[object Object]"></wbullet-stub> <span>2</span>
    </div>
  </div>
</div>
`;

exports[`Widgets/WTooltip Shouldn't be render 1`] = `undefined`;

exports[`Widgets/WTooltip Watchers Shouldn't be reset selected if changed Chart.active 1`] = `undefined`;

exports[`Widgets/WTooltip Watchers Shouldn't be set selected if changed Chart.active 1`] = `
<div style="position: absolute; display: flex; padding: .5rem; border-radius: 4px; color: rgb(0, 0, 0); background: white; box-shadow: 0 2px 10px 0 rgba(0, 0, 0, 0.15); font-size: 14px; opacity: 0; z-index: 1; transition: opacity .25s ease; top: 10px; left: 123px;">
  <div style="display: flex; flex-direction: column;"><span style="margin-bottom: 5px;">Page 1</span>
    <div><span class="WBullet" style="display: inline-block; width: 8px; height: 8px; background: white; border-radius: 50%; border-width: 2px; border-color: #000; border-style: solid; margin: 0px 8px 0px 0px;"></span> <span>Value 1</span></div>
    <div><span class="WBullet" style="display: inline-block; width: 8px; height: 8px; background: white; border-radius: 50%; border-width: 2px; border-color: #000; border-style: solid; margin: 0px 8px 0px 0px;"></span> <span>Value 2</span></div>
  </div>
</div>
`;
